### PR TITLE
[feat] Support time-based checkpointing during training

### LINKED ISSUE #6286: Support time-based checkpointing trigger
## 🚀 Feature
Support time-based checkpointing in model checkpoint callback

### Motivation
After #6146 we'll have support in Lightning to checkpoint after N training batches, or after M validation epochs. A useful feature would be to checkpoint after T time during training phase (e.g. checkpoint every 1 hour).

### Pitch

This would entail:
- adding a new optional argument `time_interval` to the callback constructor. This is of type timedelta. For all practical purposes, this cannot be smaller than the amount of time it takes to process a single training batch. This is not guaranteed to execute at the exact time specified, but should be close.
- Inside of `on_train_batch_end`: we add the following check
```
now = time.monotonic()
time_interval = self.time_interval
prev_time_check = self._prev_time_check
skip_time = (
    time_interval is None
    or prev_time_check is None
    or (now - prev_time_check) < time_interval.total_seconds()
)
if skip_batch and skip_time:
    return
if not skip_time:
    self._prev_time_check = now
...  # commence with saving checkpoint
```
note we will need a synchronization between ranks such that all ranks enter the checkpoint save logic together in case their timers are slightly off.

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #6286

To discuss: The PR currently enforces each trigger (`every_n_train_steps`, `train_time_interval`, `every_n_val_epochs`) to be mutually exclusive. This means if someone wants to checkpoint every N hours and every M train batches, they'd need to create two callbacks.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -234,8 +246,22 @@ def on_train_batch_end(
             return
         step = trainer.global_step
         skip_batch = self._every_n_train_steps < 1 or ((step + 1) % self._every_n_train_steps != 0)
-        if skip_batch:
+
+        train_time_interval = self._train_time_interval
+        skip_time = True
+        now = time.monotonic()
+        if train_time_interval is not None:
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:253]:
```suggestion
        if train_time_interval:
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -718,6 +728,41 @@ def test_ckpt_every_n_train_steps(tmpdir):
...
+                train_time_interval=timedelta(minutes=1),
+                save_top_k=-1,
+                save_last=False,
+            )
+        ],
+        logger=False,
+    )
+
+    trainer.fit(model)
+    # Each batch takes 7 sec and we checkpoint every minute. There are 64
```
Comment (Borda) [inline tests/checkpointing/test_model_checkpoint.py:760]: this is very depending on the used device, shall we rather add some sleep?
also some larger interval so with rounding, we can eliminate runtime/speed differences... 5 instead of 1

Comment (ananthsub) [reply tests/checkpointing/test_model_checkpoint.py:760]: in the test, we mock this to always take 7 seconds per batch

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -200,6 +208,7 @@ def __init__(
         self.save_weights_only = save_weights_only
         self.auto_insert_metric_name = auto_insert_metric_name
         self._last_global_step_saved = -1
+        self._prev_time_check: Optional[float]
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:211]:
```suggestion
        self._prev_time_check: Optional[float] = None
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -234,8 +246,22 @@ def on_train_batch_end(
...
+
+        train_time_interval = self._train_time_interval
+        skip_time = True
+        now = time.monotonic()
+        if train_time_interval:
+            prev_time_check = self._prev_time_check
+            skip_time = (prev_time_check is None or (now - prev_time_check) < train_time_interval.total_seconds())
+            # in case we have time differences across ranks
+            # broadcast the decision on whether to checkpoint from rank 0 to avoid possible hangs
+            skip_time = trainer.training_type_plugin.broadcast(skip_time)
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:258]: I had the same issue for the timer. it will give a perf. hit per batch

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -189,6 +196,7 @@ def __init__(
         mode: str = "min",
         auto_insert_metric_name: bool = True,
         every_n_train_steps: Optional[int] = None,
+        train_time_interval: Optional[timedelta] = None,
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:199]: it's intended for time between training_step calls. you're right that this could include validation if someone runs validation during the middle of the training epoch. i added `train` also to keep the pattern of `every_n_train_steps` / `every_n_val_epochs` which mention which running stage they're called in

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:258]: i don't see a way around it ☹️

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 6]
```diff
@@ -100,12 +102,17 @@ class ModelCheckpoint(Callback):
             is saved (``model.save(filepath)``).
         every_n_train_steps: Number of training steps between checkpoints.
             If ``every_n_train_steps == None or every_n_train_steps == 0``, we skip saving during training
-            To disable, set ``every_n_train_steps = 0``. This value must be ``None`` non-negative.
-            This must be mutually exclusive with ``every_n_val_epochs``.
+            To disable, set ``every_n_train_steps = 0``. This value must be ``None`` or non-negative.
+            This must be mutually exclusive with ``train_time_interval`` and ``every_n_val_epochs``.
+        train_time_interval: Checkpoints are monitored at the specified time interval.
+            For all practical purposes, this cannot be smaller than the amount
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:108]: if the time interval is smaller than the time to process a single batch then we won't save a checkpoint based on this trigger. we'll always skip because the time between any 2 batches will always be smaller than this interval
```
skip_time = (prev_time_check is None or (now - prev_time_check) < train_time_interval.total_seconds())
```

i dont expect anyone to actually run into this issue, but thought it was worth mentioning as a caveat

File: tests/checkpointing/test_model_checkpoint.py [hunk 7]
```diff
@@ -718,6 +728,41 @@ def test_ckpt_every_n_train_steps(tmpdir):
     assert set(os.listdir(tmpdir)) == set(expected)
 
 
+@mock.patch("pytorch_lightning.callbacks.model_checkpoint.time")
+def test_ckpt_train_time_interval(mock_datetime, tmpdir) -> None:
```
Comment (awaelchli) [inline tests/checkpointing/test_model_checkpoint.py:732]:
```suggestion
def test_model_checkpoint_train_time_interval(mock_datetime, tmpdir) -> None:
```

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
Maybe worth adding a note in the case the user wants to do multiple stuff at the same time
```
.. note:
   If you want a checkpoint at both every N hours and every M train batches or val epochs then multiple ``ModelCheckpoint`` should be created.
```
